### PR TITLE
Grocery Store - Jacquelyn Cheng - Nodes

### PR DESCRIPTION
**Pull request by jacquelynoelle**
# Grocery Store
Congratulations! You're submitting your assignment.

## Comprehension Questions

Question | Response
--- | ---
What is accomplished with `raise ArgumentError`? | It halts the program when it's provided with information it can't move forward with (e.g. invalid fulfillment status, trying to add something to an order that's already there, looking up orders for a customer that doesn't exist). |
Why do you think we made the `.all` & `.find` methods class methods?  Why not instance methods? | The .all and .find methods are class methods because they provide information that requires going through all of the Orders and Customers instantiated, not just one. |
Think about the relation between `Order` and `Customer`. Is this relation one-to-one, one-to-many, or something else? How does that compare to the Solar System project? | This is a one-to-many relationship -- a customer can have many orders but an order can belong to only one customer. This is similar to a solar system which can have many planets, but a planet can belong to only one solar system. |
How is the relation between `Order` and `Customer` tracked in the CSV file? How is it tracked in your program? Why might these be different? | In the order CSV file, orders are linked to customer ids, while in the program, they are linked to customer objects. The latter relation creates a direct link between the two objects and gives us easy access to up-to-date information about the customer. You might not want the CSV to hold extended customer information for each order, because it's static. If a customer moves, for example, a previous order might now have out-of-date customer info attached to it. This is why the CSV keeps just the id, so you must refer to the customer info for more details. |
Did the presence of automated tests change the way you thought about the problem? How? | Yes, they helped me understand how certain edge cases should be handled and what types of values should be returned by methods. They also gave me peace of mind that the program was working as intended after they all passed! |
CS Fun: If there are `C` customers listed in `customers.csv` and `R` orders listed in `orders.csv`, what is the time complexity of calling `Customer.find`? What is the time complexity of `Order.find`? | Customer.find is O(C) -- worst case scenario is that the method completes one full loop through each of the customers. Order.find is O(R) for the same reason. |

### CODE REVIEW

**Comment by dHelmgren** [comment]
# Grocery Store
## What We're Looking For

Feature | Feedback
--- | ---
**Baseline** |
Answered comprehension questions | Great answers!
Used Git Regularly | Yeah! Small commits on a regular schedule, I LOVE IT.
**Wave 1** |
All provided tests pass | Yes
Using the appropriate `attr_` for instance variables | I appreciate that you kept as much of the `Order` class as private as possible.
**Wave 2** |
All stubbed tests are implemented fully and pass | Yes indeed!
Used `CSV` library only in `.all` (not in `.find`) | See comment. I don't think calling .all in .find is ideal.
Appropriately parses the product data from CSV file in `Order.all` | Yeah!
`Order.all` calls `Customer.find` to set up the composition relation | Sho' does!
**Additional Notes** | Super solid work! Well done!

File: lib/order.rb [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+      customer = Customer.find(row[2].to_i)
+      fulfillment_status = row[3].to_sym
+      Order.new(id, products, customer, fulfillment_status)
+    end
+    return all_orders
+  end
+
+  # returns an Order instance matching the designated id
+  def self.find(order_id)
+    Order.all.each do |order|
```
Comment (dHelmgren) [inline lib/order.rb:35]: It is not find's job to call all, at least in my opinion. If you've already loaded the data, `self.find` should just find the person. If `@@orders` isn't loaded, I think it should return `nil`. This would be better separation of concerns.
